tool/tsh/common/db_exec.go

```diff
@@ -0,0 +1,626 @@
+
+func (c *sharedDatabaseExecClient) issueCert(ctx context.Context, dbInfo *databaseInfo) (tls.Certificate, error) {
+	// TODO(greedy52) add support for multi-session MFA.
+	params := client.ReissueParams{
+		RouteToCluster:  c.SiteName,
+		RouteToDatabase: client.RouteToDatabaseToProto(dbInfo.RouteToDatabase),
+		AccessRequests:  c.profile.ActiveRequests,
+	}
+
+	keyRing, _, err := c.clusterClient.IssueUserCertsWithMFA(ctx, params)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+	dbCert, err := keyRing.DBTLSCert(dbInfo.RouteToDatabase.ServiceName)
+	return dbCert, trace.Wrap(err)
+}
+
+func (c *sharedDatabaseExecClient) listDatabasesWithFilter(ctx context.Context, filter *proto.ListResourcesRequest) (databases []types.Database, err error) {
+	ctx, span := c.tracer.Start(
+		ctx,
+		"listDatabasesWithFilter",
+		oteltrace.WithSpanKind(oteltrace.SpanKindClient),
+	)
+	defer span.End()
+
+	servers, err := apiclient.GetAllResources[types.DatabaseServer](ctx, c.clusterClient.AuthClient, filter)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return types.DatabaseServers(servers).ToDatabases(), nil
+}
+
+type databaseExecCommandMaker struct {
+	tc          *client.TeleportClient
+	profile     *client.ProfileStatus
+	rootCluster string
+}
+
+func newDatabaseExecCommandMaker(ctx context.Context, tc *client.TeleportClient, profile *client.ProfileStatus) (*databaseExecCommandMaker, error) {
+	rootCluster, err := tc.RootClusterName(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &databaseExecCommandMaker{
+		tc:          tc,
+		profile:     profile,
+		rootCluster: rootCluster,
+	}, nil
+}
+
+func (m *databaseExecCommandMaker) makeCommand(ctx context.Context, dbInfo *databaseInfo, lpAddr, command string) (*exec.Cmd, error) {
+	addr, err := utils.ParseAddr(lpAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	opts, err := makeDatabaseCommandOptions(ctx, m.tc, dbInfo,
+		dbcmd.WithLocalProxy("localhost", addr.Port(0), ""),
+		dbcmd.WithNoTLS(),
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return dbcmd.NewCmdBuilder(m.tc, m.profile, dbInfo.RouteToDatabase, m.rootCluster, opts...).
+		GetExecCommand(ctx, command)
+}
+
+// ensureEachDatabase ensures one to one mapping between the provided database
+// target names and database resources.
+//
+// Note that it is assumed that the provided database resource has at least one
+// matching names as they are retrieved from the backend based on one of those
+// names.
+func ensureEachDatabase(names []string, dbs []types.Database) error {
+	byDiscoveredNameOrName := map[string]types.Databases{}
+	for _, db := range dbs {
+		byDiscoveredNameOrName[db.GetName()] = append(byDiscoveredNameOrName[db.GetName()], db)
+
+		// Database may be listed by their original name in the cloud.
+		if discoveredName, ok := common.GetDiscoveredResourceName(db); ok {
+			byDiscoveredNameOrName[discoveredName] = append(byDiscoveredNameOrName[discoveredName], db)
+		}
+	}
```

Comment:
I think that as written, this function is subtly wrong, for two reasons:
1. If "discovered name" exists it may be identical to `db.GetName()` and we end up adding the db twice.
2. Unlike the predicate construction, here we don't trim the whitespace around name so we may miss a match.

Reply:
fixed 1 here. for 2, i decided to do the trim space earlier when splitting the flag.